File: modules/Btag_weight.cc
#include <iostream>
#include <TH1.h>
#include <string>
#include <TEfficiency.h>
#include <ROOT/RVec.hxx>
#include "correction.h"

using namespace ROOT::VecOps;

float evaluate_corr(std::shared_ptr<const correction::Correction> bc_sf, std::shared_ptr<const correction::Correction> incl_sf, TEfficiency* eff_b, TEfficiency* eff_c, TEfficiency* eff_light, RVec<ROOT::Math::PtEtaPhiMVector> Jet_vec, RVec<int> Jet_hadronFlavour, RVec<float> Jet_btagDisc, std::string wp, float wp_val) {
    // Calculate nominal event reweight - handles scale factors for b/c/light jets in one function
    float val_tag   = 1.0;
    float val_notag = 1.0;
    for (size_t ijet = 0; ijet < Jet_vec.size(); ijet++) {
        // find MC efficiency/scale factor
	float efficiency;
	float sf;
	float pt = Jet_vec[ijet].Pt();
	float abseta = std::abs(Jet_vec[ijet].Eta());
        if (Jet_hadronFlavour[ijet] == 5) { // b-jet
	    int globalbin = eff_b->FindFixBin(pt,abseta);
	    efficiency = eff_b->GetEfficiency(globalbin); 
    	    sf = (float)bc_sf->evaluate({"central", wp.c_str(), Jet_hadronFlavour[ijet], abseta, pt});
        }
	else if (Jet_hadronFlavour[ijet] == 4) { // c-jet
            int globalbin = eff_c->FindFixBin(pt,abseta);
            efficiency = eff_c->GetEfficiency(globalbin);
            sf = (float)bc_sf->evaluate({"central", wp.c_str(), Jet_hadronFlavour[ijet], abseta, pt});
        } 
	else { // light jet
	    int globalbin = eff_light->FindFixBin(pt,abseta);
            efficiency = eff_light->GetEfficiency(globalbin);
            sf = (float)incl_sf->evaluate({"central", wp.c_str(), Jet_hadronFlavour[ijet], abseta, pt});
	}

        // Calculate contribution to event weight
        if (Jet_btagDisc[ijet] > wp_val) {
	    float valTag = (sf * efficiency) / (efficiency);
            val_tag *= valTag;
        }
	else {
            float valNoTag = (1. - sf*efficiency) / (1. - efficiency);
            val_notag *= valNoTag;
	}
    }
    return val_tag * val_notag; // overall event weight
}

RVec<float> evaluate_uncert(std::shared_ptr<const correction::Correction> sf_set, std::string corr, std::vector<int> hadron_flavours, std::vector<TEfficiency*> eff_maps, RVec<ROOT::Math::PtEtaPhiMVector> Jet_vec, RVec<int> Jet_hadronFlavour, RVec<float> Jet_btagDisc, std::string wp, float wp_val) {
    float val_tag_up   = 1.0;
    float val_notag_up = 1.0;
    float val_tag_dn   = 1.0;
    float val_notag_dn = 1.0;
    // Loop over all jets 
    for (size_t ijet = 0; ijet < Jet_vec.size(); ijet++) {
	// Check if jet is one of the target flavors
	auto it = std::find(hadron_flavours.begin(),hadron_flavours.end(),Jet_hadronFlavour[ijet]);
	if (it != hadron_flavours.end()) {
	    int fidx = it - hadron_flavours.begin();
            float pt = Jet_vec[ijet].Pt();
            float abseta = std::abs(Jet_vec[ijet].Eta());
	    // get efficiency/scale factors
	    int globalbin = eff_maps[fidx]->FindFixBin(pt,abseta);
            float efficiency = eff_maps[fidx]->GetEfficiency(globalbin);
            float sf_up = (float)sf_set->evaluate({"up_"+corr, wp.c_str(), Jet_hadronFlavour[ijet], abseta, pt});
            float sf_down = (float)sf_set->evaluate({"down_"+corr, wp.c_str(), Jet_hadronFlavour[ijet], abseta, pt});
        
	    // calculate contribution to event weight
	    if (Jet_btagDisc[ijet] > wp_val) {
                float valUpTag = (sf_up * efficiency) / (efficiency);
                float valDnTag = (sf_down * efficiency) / (efficiency);
                val_tag_up *= valUpTag;
                val_tag_dn *= valDnTag;
            }
            else {
                float valUpNoTag = (1. - (sf_up * efficiency)) / (1. - efficiency);
                float valDnNoTag = (1. - (sf_down * efficiency)) / (1. - efficiency);
                val_notag_up *= valUpNoTag;
                val_notag_dn *= valDnNoTag;
            }
	}
    }
    float weight_up = val_tag_up * val_notag_up;
    float weight_dn = val_tag_dn * val_notag_dn;
    return {weight_up, weight_dn};
}

File: modules/Btag_weight.py
'''
Script to run b-tagging corrections in timber with correctionlib 
'''
import ROOT
from TIMBER.Analyzer import *
from TIMBER.Tools.Common import *
import correctionlib

def BTagSF(self,jet_coll,wp='T',mujets_or_comb='comb'):
    # self: XHYbbWW object on which to perform SF application 
    # jet_coll: name of jet collection to consider for b-tagging (i.e. "Jet", "kinJet",etc...) - this constitues all jets for which we query the b-tagging status in the analysis
    # wp: working point for tagger used ('L','M', or 'T')
    # mujets_or_comb: choose from "mujets" or "comb" variant of SF's for b/c jets

    # need to do this for it to work
    correctionlib.register_pyroot_binding() 

    # Compile the C++ functions/classes we need 
    CompileCpp('modules/Btag_weight.cc') 

    # load the btagging correction set
    print('Loading BTV SF correction set')
    if '16' in self.year:
        if self.year == '16APV' : cyear = '16preVFP'
        else: cyear = '16postVFP'
    else:
        cyear = self.year
    ROOT.gInterpreter.Declare(f'auto btag_sf_set = correction::CorrectionSet::from_file("/cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/BTV/20{cyear}_UL/btagging.json.gz");')

    # For DeepJet we use the following scale factors:
    #   - SFbc:
    #       * either "mujets" or "comb" SFs are to be applied 
    #         mujets: SFs derived from QCD-enriched regions; 
    #         comb: additionally also SFs from ttbar enriched regions
    #   - SFlight:
    #       * for light jets (hadron flavor 0) "incl" SFs
    # One common multiplicative SF is to be applied to the nominal prediction.
    # Separate uncertainties are applied for b/c jets and light jets.

    # Get all of the relevant SFs
    ROOT.gInterpreter.Declare(f'auto deepJet_bc_sf = btag_sf_set->at("deepJet_{mujets_or_comb}");') # b/c SFs 
    ROOT.gInterpreter.Declare('auto deepJet_incl_sf = btag_sf_set->at("deepJet_incl");')     # light SFs

    '''
    # Rebin efficiency map
    rebinFactor = 6 if 'XHY' in self.setname else 2
    effFile = ROOT.TFile.Open(f"{path}{self.setname}_{self.year}_bJets_btagEfficiencies.root")
    hAll = effFile.Get('all').RebinY(rebinFactor)
    hTagged = effFile.Get('tagged').RebinY(rebinFactor)
    eff = ROOT.TEfficiency(hTagged,hAll)
    eff.SetName('eff_rebin')
    eff.Write()
    effFile.Close()
    '''

    # Setnames for efficiency maps 
    combined_procs = ['DYJets','ST','WJetsHT','WJetsLepHT','ZJets']
    effname = self.setname
    for name in combined_procs:
        if name in self.setname:
            effname = name
            break


    # Load efficiency maps
    print("Loading b-tagging efficiency maps")
    path = 'root://cmseos.fnal.gov//store/user/mhesford/XHYbbWW_semileptonic/plots/btagEfficiencies/'
    print(f'{path}{effname}_{self.year}_bJets_btagEfficiencies.root')
    print(f'{path}{effname}_{self.year}_cJets_btagEfficiencies.root')
    print(f'{path}{effname}_{self.year}_lightJets_btagEfficiencies.root')
    ROOT.gInterpreter.Declare(f'TEfficiency* eff_b = (TEfficiency*)TFile::Open("{path}{effname}_{self.year}_bJets_btagEfficiencies.root")->Get("eff");')
    ROOT.gInterpreter.Declare(f'TEfficiency* eff_c = (TEfficiency*)TFile::Open("{path}{effname}_{self.year}_cJets_btagEfficiencies.root")->Get("eff");')
    ROOT.gInterpreter.Declare(f'TEfficiency* eff_light = (TEfficiency*)TFile::Open("{path}{effname}_{self.year}_lightJets_btagEfficiencies.root")->Get("eff");')

    # Value to cut on for b-tagging
    wp_val = self.cuts['JET']['btagDeepJet'][wp][self.year]

    # Evaluate nominal b-tagging event weights using method 1A described here:
    # https://btv-wiki.docs.cern.ch/PerformanceCalibration/fixedWPSFRecommendations/#scale-factor-recommendations-for-event-reweighting
    self.a.Define('Btag_weight_nom',f'evaluate_corr(deepJet_bc_sf,deepJet_incl_sf,eff_b,eff_c,eff_light,{jet_coll}_vect,{jet_coll}_hadronFlavour,{jet_coll}_btagDeepFlavB,"{wp}",{wp_val})')
    #self.a.DataFrame.Display([f'Btag_weight_nom']).Print()  # debug

    #Evaluate up/down uncertainty weights
    # A breakdown of full Run2 uncertainties can be found here: 
    # https://btv-wiki.docs.cern.ch/PerformanceCalibration/SFUncertaintiesAndCorrelations/#working-point-based-sfs-fixedwp-sfs
    for corr in ['correlated','uncorrelated']: 
        # Apply a fully correlated uncertainty across the four data taking eras and a separate uncorrelated one
        self.a.Define(f'Btag_weight_uncert_bc_{corr}',f'evaluate_uncert(deepJet_bc_sf,"{corr}",{{5,4}},{{eff_b,eff_c}},{jet_coll}_vect,{jet_coll}_hadronFlavour,{jet_coll}_btagDeepFlavB,"{wp}",{wp_val})') #b/c
        self.a.Define(f'Btag_weight_uncert_light_{corr}',f'evaluate_uncert(deepJet_incl_sf,"{corr}",{{0}},{{eff_light}},{jet_coll}_vect,{jet_coll}_hadronFlavour,{jet_coll}_btagDeepFlavB,"{wp}",{wp_val})') #light
        
        #self.a.DataFrame.Display([f'Btag_weight_uncert_bc_{corr}']).Print()  # debug
        #self.a.DataFrame.Display([f'Btag_weight_uncert_light_{corr}']).Print()  # debug

    # Now use the TIMBER BranchCorrection module to add the weight columns as corrections to be tracked
    btag_weight_nom = Correction('Btag','TIMBER/Framework/src/BranchCorrection.cc',corrtype='corr',mainFunc="evalCorrection") #nominal correction
    self.a.AddCorrection(btag_weight_nom, evalArgs={'val':'Btag_weight_nom'}) 
    # Up/down uncertainties
    for corr in ['correlated','uncorrelated']:
        for bc_l in ['bc','light']:
            btag_weight_uncert = btag_weight_nom.Clone(f'Btag_{bc_l}_{corr}',newMainFunc="evalUncert",newType="uncert")
            self.a.AddCorrection(btag_weight_uncert, evalArgs={'valUp':f'Btag_weight_uncert_{bc_l}_{corr}[0]','valDown':f'Btag_weight_uncert_{bc_l}_{corr}[1]'})

